crates/core/src/topology.rs

```diff
@@ -39,49 +47,62 @@ const RANDOM_CLOSEST_DISTANCE: f64 = 1.0 / 1000.0;
 /// time window. The goal of this is to select the best connections over time
 /// from incoming join requests.
 pub(crate) struct TopologyManager {
-    slow_connection_evaluator: connection_evaluator::ConnectionEvaluator,
-    fast_connection_evaluator: connection_evaluator::ConnectionEvaluator,
-    request_density_tracker: request_density_tracker::RequestDensityTracker,
+    limits: Limits,
+    meter: Meter,
+    source_creation_times: DashMap<AttributionSource, Instant>,
```

Comment:
I am not sure why this needs to be a concurrent data structure since access to topology manager itself is always behing a rwlock already, a regular HashMap hould be enough here

Reply:
You're right, converted to HashMap.